File: src/backend/src/websocket/ws.js
const WebSocket = require("ws");
const mqttService = require("../services/mqtt.service");
const config = require("../config/env");

class WebSocketServer {
  constructor() {
    this.wss = null;
    this.clients = new Set();
  }

  initialize(server) {
    this.wss = new WebSocket.Server({ server });

    this.wss.on("connection", (ws) => {
      console.log("✓ New WebSocket client connected");
      this.clients.add(ws);

      ws.on("message", (message) => {
        try {
          const data = JSON.parse(message);
          console.log("WebSocket message received:", data);

          // Handle client requests
          this.handleClientMessage(ws, data);
        } catch (error) {
          console.error("Error parsing WebSocket message:", error);
        }
      });

      ws.on("close", () => {
        console.log("WebSocket client disconnected");
        this.clients.delete(ws);
      });

      ws.on("error", (error) => {
        console.error("WebSocket error:", error);
        this.clients.delete(ws);
      });

      // Send welcome message
      ws.send(
        JSON.stringify({
          type: "connection",
          message: "Connected to IoT Backend WebSocket",
          timestamp: new Date().toISOString(),
        })
      );
    });

    // Forward MQTT messages to all WebSocket clients
    mqttService.onMessage((topic, payload) => {
      this.broadcast({
        type: "sensor_data",
        topic,
        data: payload,
        timestamp: new Date().toISOString(),
      });
    });

    console.log(`✓ WebSocket server initialized`);
  }

  handleClientMessage(ws, data) {
    switch (data.type) {
      case "ping":
        ws.send(
          JSON.stringify({ type: "pong", timestamp: new Date().toISOString() })
        );
        break;
      case "subscribe":
        // Handle subscription requests
        ws.send(
          JSON.stringify({
            type: "subscribed",
            topic: data.topic,
            timestamp: new Date().toISOString(),
          })
        );
        break;
      default:
        ws.send(
          JSON.stringify({
            type: "error",
            message: "Unknown message type",
            timestamp: new Date().toISOString(),
          })
        );
    }
  }

  broadcast(data) {
    const message = JSON.stringify(data);
    this.clients.forEach((client) => {
      if (client.readyState === WebSocket.OPEN) {
        client.send(message);
      }
    });
  }

  close() {
    if (this.wss) {
      this.wss.close();
      console.log("WebSocket server closed");
    }
  }
}

module.exports = new WebSocketServer();

File: src/backend/src/websocket/ws.py
from fastapi import WebSocket, WebSocketDisconnect
from typing import Set, Dict, Any
import json
from datetime import datetime
from app.services.mqtt_service import mqtt_service


class WebSocketManager:
    def __init__(self):
        self.clients: Set[WebSocket] = set()

    async def connect(self, websocket: WebSocket):
        """Accept new WebSocket connection"""
        await websocket.accept()
        self.clients.add(websocket)
        print(f"✓ New WebSocket client connected (total: {len(self.clients)})")

        # Send welcome message
        await websocket.send_json({
            "type": "connection",
            "message": "Connected to IoT Backend WebSocket",
            "timestamp": datetime.now().isoformat()
        })

    def disconnect(self, websocket: WebSocket):
        """Remove WebSocket connection"""
        self.clients.discard(websocket)
        print(f"WebSocket client disconnected (remaining: {len(self.clients)})")

    async def handle_message(self, websocket: WebSocket, data: Dict[str, Any]):
        """Handle incoming WebSocket messages"""
        msg_type = data.get("type")

        if msg_type == "ping":
            await websocket.send_json({
                "type": "pong",
                "timestamp": datetime.now().isoformat()
            })
        elif msg_type == "subscribe":
            await websocket.send_json({
                "type": "subscribed",
                "topic": data.get("topic"),
                "timestamp": datetime.now().isoformat()
            })
        else:
            await websocket.send_json({
                "type": "error",
                "message": "Unknown message type",
                "timestamp": datetime.now().isoformat()
            })

    async def broadcast(self, data: Dict[str, Any]):
        """Broadcast message to all connected clients"""
        message = json.dumps(data)
        disconnected = set()

        for client in self.clients:
            try:
                await client.send_text(message)
            except Exception as e:
                print(f"Error sending to client: {e}")
                disconnected.add(client)

        # Remove disconnected clients
        for client in disconnected:
            self.clients.discard(client)

    def initialize(self):
        """Initialize WebSocket manager and MQTT forwarding"""
        # Forward MQTT messages to all WebSocket clients
        def mqtt_to_ws(topic: str, payload: Dict[str, Any]):
            import asyncio
            asyncio.create_task(self.broadcast({
                "type": "sensor_data",
                "topic": topic,
                "data": payload,
                "timestamp": datetime.now().isoformat()
            }))

        mqtt_service.on_message(mqtt_to_ws)
        print("✓ WebSocket manager initialized")


# Singleton instance
ws_manager = WebSocketManager()